firedrake/cython/dmcommon.pyx

```diff
@@ -822,12 +1009,20 @@ def get_cell_nodes(mesh,
                 # we need to offset by the difference from the bottom.
                 if variable:
                     off += offset[flat_index[k]]*(layer_extents[c, 0] - layer_extents[entity, 0])
-                for j in range(ceil_ndofs[i]):
-                    cell_nodes[cell, flat_index[k]] = off + j
-                    k += 1
+                if entity_permutations is not None:
+                    for j in range(ceil_ndofs[i]):
+                        cell_nodes[cell, flat_index[k]] = off + entity_permutations[entity_dims[i]][entity_nums[i]][orient][j]
```

Comment:
I think this is "python indexing into a dict" in an inner loop. Can you provide short benchmarking of how much difference it makes to timing for (say) a `50x50x50` mesh with P4?

Reply:
original               : 0.73 min
with permutation: 2.85 min
Indeed very slow.

Reply:
Great improvement with `entity_permutations` preprocessed:

50x50x50 P4 example:

off + j                                                 : 0.143 min
off + perm[j] (without preprocessing): 1.47 min
off + perm[j]  (with preprocessing)     : 0.177 min

(I probably used different computer in my earlier analysis, so values are different).